awx/main/scheduler/dag_workflow.py

```diff
@@ -97,3 +100,41 @@ def is_workflow_done(self):
                 # have the job result.
                 return False, False
         return True, is_failed
+
+    def mark_dnr_nodes(self):
+        root_nodes = self.get_root_nodes()
+        nodes = copy.copy(root_nodes)
+        nodes_marked_do_not_run = []
+
+        for index, n in enumerate(nodes):
+            obj = n['node_object']
+            job = obj.job
+
+            if not job and obj.do_not_run is False and n not in root_nodes:
+                parent_nodes = [p['node_object'] for p in self.get_dependents(obj)]
+                all_parents_dnr = True
+                for p in parent_nodes:
+                    if not p.job and p.do_not_run is False:
+                        all_parents_dnr = False
+                        break
+                #all_parents_dnr = reduce(lambda p: bool(p.do_not_run == True), parent_nodes)
+                if all_parents_dnr:
+                    obj.do_not_run = True
+                    nodes_marked_do_not_run.append(n)
+
+            if obj.do_not_run:
+                children_success = self.get_dependencies(obj, 'success_nodes')
+                children_failed = self.get_dependencies(obj, 'failure_nodes')
+                children_always = self.get_dependencies(obj, 'always_nodes')
+                children_all = children_failed + children_always
+                nodes.extend(children_all)
```

Comment:
Since the network now has convergence, elements in `children_all` may already exist in `nodes`. If that's the case, then I don't think they should be added to the list. It will still "work" to do this, but dense networks will traverse all paths, which can be inefficient to a exponential degree.